### ISSUE TITLE
Foreign content attributes do not serialize properly

### ISSUE THREAD

**Issue by gsnedders** [html5lib/html5lib-python#59]
```
In [6]: html5lib.serialize(html5lib.parse("<math><mi xml:lang=en xlink:href=foo /></math>"))
Out[6]: u'<math><mi href=foo lang=en></mi></math>'
```

We should get `u'<math><mi xml:href=foo xml:lang=en></mi></math>'` or similar back.

**Comment by willkg** [comment]
Pretty sure I just hit this. I don't know if it's the same cause as it was in 2013, but I'm pretty sure the problem is here:

https://github.com/html5lib/html5lib-python/blob/17499b9763a090f7715af49555d21fe4b558958b/html5lib/serializer.py#L246

Given that there's a `# TODO` there, I think it's probably known. Figured I'd mention it in the bug comments in case that's helpful.

**Comment by gsnedders** [comment]
Yeah, this is something I keep saying I'll fix, but IIRC it's somewhat awkward for some reason that's never quite obvious.

**Comment by frivoal** [comment]
hi @gsnedders 

I think I just hit that too, and it's getting in the way of publishing specifications at W3C using echidna:
1. Bikeshed generates a railroad diagram using `<a xlink:href="..."` in svg
2. html5lib-python is used by bikeshed to process the markup, and strips the `xlink:` namespace, leaving `<a href="..."
3. try and publish the spec using echidna
4. echidna uses validator.nu, which doesn't allow href without `xlink:` in svg, and rejects the spec
5. ...
6. No profit

Sure, echidna/validator.nu shouldn't reject that(https://github.com/validator/validator/issues/511), but they do, and I cannot work around the problem by asking bikeshed do generate links with `xlink:` since it already does before html5lib strips it out.

Can I haz fix plz?

cc: @tabatkins

**Comment by gsnedders** [comment]
I thought Bikeshed had its own HTML serializer?

**Comment by frivoal** [comment]
I don't know how bikeshed works, but it invokes html5lib at least for some things, behaves the same as if it was triggering this bug, and if I change bikeshed to emit an xlink:hrof instead of xlink:href just for testing purposes, it throws this at me:
> /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/html5lib/_ihatexml.py:265: DataLossWarning: Coercing non-XML name
  warnings.warn("Coercing non-XML name", DataLossWarning)

which I don't quite understand or want to debug, but having html5lib in the name makes me think html5lib is invoked by bikeshed for some kind of post processing, and that I am hitting this bug.

Anyway, @sideshowbarker did fix validator.nu, so this is no longer blocking me, but it probably would still be good to fix.

**Comment by tabatkins** [comment]
Yeah, Bikeshed uses its own custom serializer. That warning from html5lib is on the *parsing* side, and it's because the HTML parser only recognizes a handful of "namespaced" attributes, and if you do something else you just get an attribute name with a colon in it, which is illegal per XML rules.

(I hate *hate* ***hate*** that particular error, as it gives *zero* information about what is being coerced, for no good reason. It's usually caused by forgetting to put a closing quote on an attribute value, so you end up with weird garbage. I've complained about this issue before to @gsnedders.)

**Comment by gsnedders** [comment]
(@tabatkins, can you file a bug on that?)

**Comment by tabatkins** [comment]
I did. ^_^ #107
